[Tests] Re-enable the tests for integer overflow diagnostics

ravikandhadai opened the pull request:
This commit splits up the test suite 'diagnostic_constant_propagation_propagation_int.swift'
into two files one each for 64bit and 32bit architectures.
The tests check the diagnostics for arithmetic and bitwise operations on
'Int' and 'UInt' values. The size of these numeric types depend on the word length of
the target architecture. Most of the tests added in this commit were present
earlier but were xfailed.

ravikandhadai committed:
[Tests] Re-enable the tests for integer overflow diagnostics

This commit splits up the test suite 'diagnostic_constant_propagation_propagation_int.swift'
into two files one each for 64bit and 32bit architectures.
The tests check the diagnostics for arithmetic and bitwise operations on
'Int' and 'UInt' values. The size of these numeric types depend on the word length of
the target architecture. Most of the tests added in this commit were present
earlier but were xfailed.

devincoughlin commented on test/SILOptimizer/diagnostic_constant_propagation_int_arch64.swift:
```diff
@@ -0,0 +1,235 @@
+// RUN: %target-swift-frontend -emit-sil -primary-file %s -o /dev/null -verify
+//
+// REQUIRES: CPU=x86_64 || CPU=arm64 || CPU=powerpc64 || CPU=powerpc64le || CPU=s390x
```
I wonder it might be better to use PTRSIZE=64 and PTRSIZE=32. Pointer size and Int size aren't the same thing -- but this would avoid having to enumerate all 64-bit CPUs

devincoughlin commented on test/SILOptimizer/diagnostic_constant_propagation_int_arch64.swift:
```diff
@@ -0,0 +1,235 @@
+  }
+
+  do {
+    // bitwise operations. No overflows canhappen during these operations
```
Typo "canhappen"

devincoughlin reviewed (approved):
Looks good to me. It is great to see these re-enabled and using -verify instead of FileCheck!

See one comment inline about using PTRSIZE instead of enumerating architectures.

moiseev commented on test/SILOptimizer/diagnostic_constant_propagation_int_arch32.swift:
```diff
@@ -0,0 +1,232 @@
+// RUN: %target-swift-frontend -emit-sil -primary-file %s -o /dev/null -verify
+//
+// REQUIRES: CPU=i386 || CPU=arm
```
You can do `||` in `REQUIRES`?! o_O

moiseev reviewed (approved):
:shipit:

ravikandhadai commented:
Hi Devin and Max,
Thanks for the comments. I have changed the requires to use PTRSIZE and it seems to work on my system (64bit) and iPhone-i386-simulator. So I have used it instead of checking for specific architectures.

ravikandhadai commented:
Please smoke test merge

ravikandhadai commented:
@swift-ci Please smoke test merge

moiseev commented:
@ravikandhadai this is not the real command =)

moiseev commented:
@swift-ci Please smoke test

ravikandhadai commented:
@swift-ci  Please smoke test merge

ravikandhadai committed:
[Tests] Fix Radar problem <rdar:/19508336>
"Extend test/SILPasses/diagnostic_constant_propagation.swift" by removing
tests from the file that referred to `Int` and `UInt` types.
These tests are dependent on the word length of the architecture and
are covered by the tests in the files
`diagnostic_constant_propagation_int_archXX.swift`.

Also, this commit corrects erroneous references to a radar problem
in the test suties: `diagnostic_constant_propagation_int_archXX.swift`

moiseev commented:
@swift-ci Please smoke test

moiseev commented:
@swift-ci Please smoke test

xedin commented:
@devincoughlin @moiseev @ravikandhadai Would it be possible to add tests with more complex expressions as well to check that this is not too eager? E.g.

```swift
_ = Int16(integerLiteral: 0).magnitude

error: arithmetic operation '65535 + 1' (on unsigned 16-bit integer type) results in an overflow
Int16(integerLiteral: 0).magnitude
                         ^
```

The problem here is that getter for the `magnitude` uses ternary operator and one of the branches would become `~0 + 1` which is diagnosed as an overflow here incorrectly.

ravikandhadai commented:
@xedin Is the branch containing ~0 + 1 actually dead branch that will not be taken?  let's catch up to discuss more on this.

xedin commented:
@ravikandhadai yes, in this case `Int16(integerLiteral: 0)` that brach is dead because the check is `self < (0 as ${Self}) ? ~base + 1 : base` here is the actual code https://github.com/apple/swift/blob/master/stdlib/public/core/Integers.swift.gyb#L3529

xedin commented:
https://bugs.swift.org/browse/SR-6602
